### ISSUE TITLE
Release the GIL on CPU intensive methods

### ISSUE THREAD

**Issue by AntoinePrv** [scipopt/PySCIPOpt#430]
The Python [Global Interpreter Lock](https://wiki.python.org/moin/GlobalInterpreterLock) (GIL) only lets one thread execute Python code at the time.
So when using `threading` with pure Python, the code is actually concurrent, but not parallel.

When running non Python code, such as the C code of libscip, the GIL can be realsed, allowing `threading` to have some level of parallelism.

In Cython, this can be done by wrapping C function calls with:
```Cython
with nogil:
    SCIPsolve(self._scip)
```
However, one needs to make sure that no Python code is executed when releasing the GIL. This can be hard for two reasons:
  - Cython generates some Python code which can be hard to distinguish from C code;
  - The GIL needs to be reacquired in all Python callbacks such as BranchRule

**Comment by fserra** [comment]
But scip itself is sequential. Maybe this would be helpful for concurrent scip, or for the parallelism that @stephenjmaher implemented.

**Comment by AntoinePrv** [comment]
This is not about SCIP parallelism because Python would not care about C threads that do not run Python code.
This is about parallelism of user code, such as solving multiple Scip models in parallel (perhaps need `PARA_SCIP=ON`), or more generally doing any other non-SCIP computation while SCIP is solving

**Comment by fserra** [comment]
ah ok, I see. Well, I don't know anything about this. I cannot help, sorry

**Comment by AntoinePrv** [comment]
I can send a PR in the future. Just wanted to open an issue to keep track of it in the meantime.

**Comment by AntoinePrv** [comment]
@CGraczyk, in case you are still interested in this feature, #505 is not a (full) fix for this.

**Comment by CGraczyk** [comment]
I see, then i will reopen the issue, if you think this is nice to have feel free to open a new pull request.

**Comment by AntoinePrv** [comment]
@axelvonkamp, as you noticed, releasing the GIL in PySCIPOpt is not extremely hard, simply adding a number of `nogil` around (the GIL is already being re-aquired during Python callbacks).
For it to get into PySCIPOpt, it's mostly a matter of doing and testing it properly, but I personally don't have much time for it.

As far as SCIP parallelism is involved, I'm not sure when it safe, I believe compiling SCIP with `-D PARASCIP=ON` (it's already the case if you get SCIP/PySCIPOpt from `conda`) should be enough if you use separate Models in each threads (as opposed to multiple threads modifying a single Model).

**Comment by Joao-Dionisio** [comment]
Is this still an issue, @AntoinePrv?
